File: core/media/src/main/kotlin/com/echo/core/media/tracking/PlayTrackingService.kt
package com.echo.core.media.tracking

import android.util.Log
import com.echo.core.datastore.preferences.ServerPreferences
import com.echo.core.network.api.ApiClientFactory
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Service for tracking play events and updating playback state.
 * Integrates with the backend for analytics and social "listening now" feature.
 */
@Singleton
class PlayTrackingService @Inject constructor(
    private val apiClientFactory: ApiClientFactory,
    private val serverPreferences: ServerPreferences
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private var currentTrackId: String? = null
    private var currentContext: PlayContext = PlayContext.DIRECT
    private var currentSourceId: String? = null
    private var currentSourceType: SourceType? = null
    private var playStartTime: Long = 0
    private var lastPositionMs: Long = 0

    private suspend fun getApi(): PlayTrackingApi? {
        return try {
            val server = serverPreferences.activeServer.first() ?: return null
            apiClientFactory.getClient(server.url).create(PlayTrackingApi::class.java)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get API", e)
            null
        }
    }

    /**
     * Set the play context for the current/next track
     */
    fun setPlayContext(
        context: PlayContext,
        sourceId: String? = null,
        sourceType: SourceType? = null
    ) {
        currentContext = context
        currentSourceId = sourceId
        currentSourceType = sourceType
    }

    /**
     * Called when a new track starts playing
     */
    fun onTrackStarted(trackId: String) {
        // Record previous track if there was one playing
        if (currentTrackId != null && currentTrackId != trackId) {
            val listenedMs = lastPositionMs
            val wasPlaying = playStartTime > 0
            if (wasPlaying && listenedMs > 0) {
                // Record the previous track's play
                recordPlayInternal(currentTrackId!!, listenedMs)
            }
        }

        currentTrackId = trackId
        playStartTime = System.currentTimeMillis()
        lastPositionMs = 0

        // Update playback state for social "listening now"
        updatePlaybackStateInternal(isPlaying = true, trackId = trackId)
    }

    /**
     * Called when playback position changes
     */
    fun onPositionChanged(positionMs: Long) {
        lastPositionMs = positionMs
    }

    /**
     * Called when track completes
     */
    fun onTrackCompleted(trackId: String, durationMs: Long) {
        if (trackId == currentTrackId) {
            recordPlayInternal(trackId, durationMs, completionRate = 1.0f)
            currentTrackId = null
            playStartTime = 0
            lastPositionMs = 0
        }
    }

    /**
     * Called when user skips to next track
     */
    fun onTrackSkipped(trackId: String, positionMs: Long, durationMs: Long) {
        if (trackId == currentTrackId && durationMs > 0) {
            val completionRate = positionMs.toFloat() / durationMs.toFloat()

            scope.launch {
                try {
                    getApi()?.recordSkip(
                        RecordSkipData(
                            trackId = trackId,
                            playContext = currentContext.toApiValue(),
                            completionRate = completionRate,
                            sourceId = currentSourceId,
                            sourceType = currentSourceType?.toApiValue()
                        )
                    )
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to record skip", e)
                }
            }
        }
    }

    /**
     * Called when playback is paused
     */
    fun onPlaybackPaused() {
        updatePlaybackStateInternal(isPlaying = false, trackId = currentTrackId)
    }

    /**
     * Called when playback is resumed
     */
    fun onPlaybackResumed() {
        updatePlaybackStateInternal(isPlaying = true, trackId = currentTrackId)
    }

    /**
     * Called when playback is stopped
     */
    fun onPlaybackStopped() {
        // Record partial play if there was one
        if (currentTrackId != null && lastPositionMs > 0) {
            recordPlayInternal(currentTrackId!!, lastPositionMs)
        }

        updatePlaybackStateInternal(isPlaying = false, trackId = null)

        currentTrackId = null
        playStartTime = 0
        lastPositionMs = 0
    }

    private fun recordPlayInternal(
        trackId: String,
        listenedMs: Long,
        completionRate: Float? = null
    ) {
        // Only record if listened for at least 10 seconds
        if (listenedMs < 10_000) return

        scope.launch {
            try {
                getApi()?.recordPlay(
                    RecordPlayData(
                        trackId = trackId,
                        playContext = currentContext.toApiValue(),
                        completionRate = completionRate,
                        sourceId = currentSourceId,
                        sourceType = currentSourceType?.toApiValue()
                    )
                )
            } catch (e: Exception) {
                Log.e(TAG, "Failed to record play", e)
            }
        }
    }

    private fun updatePlaybackStateInternal(isPlaying: Boolean, trackId: String?) {
        scope.launch {
            try {
                getApi()?.updatePlaybackState(
                    PlaybackStateData(
                        isPlaying = isPlaying,
                        currentTrackId = trackId
                    )
                )
            } catch (e: Exception) {
                Log.e(TAG, "Failed to update playback state", e)
            }
        }
    }

    /**
     * Get play history
     */
    suspend fun getPlayHistory(limit: Int = 50): Result<List<PlayHistoryEntry>> = runCatching {
        getApi()?.getPlayHistory(limit) ?: emptyList()
    }

    /**
     * Get top tracks
     */
    suspend fun getTopTracks(
        limit: Int = 20,
        timeRange: String = "all"
    ): Result<List<TopTrack>> = runCatching {
        getApi()?.getTopTracks(limit, timeRange) ?: emptyList()
    }

    /**
     * Get recently played
     */
    suspend fun getRecentlyPlayed(limit: Int = 20): Result<List<RecentlyPlayed>> = runCatching {
        getApi()?.getRecentlyPlayed(limit) ?: emptyList()
    }

    /**
     * Get play summary
     */
    suspend fun getPlaySummary(): Result<PlaySummary> = runCatching {
        getApi()?.getPlaySummary() ?: throw IllegalStateException("No server connection")
    }

    companion object {
        private const val TAG = "PlayTrackingService"
    }
}

File: web/src/shared/services/play-tracking.service.ts
/**
 * Play Tracking Service
 *
 * Handles communication with the play tracking API endpoints.
 * Tracks user play behavior including context, completion rates, and skip events.
 *
 * @module play-tracking.service
 */

import { apiClient } from './api';
import { logger } from '@shared/utils/logger';

/**
 * Play context types
 * Determines how a track was initiated for weighted scoring
 */
export type PlayContext =
  | 'direct'         // Direct play from track list (weight: 1.0)
  | 'search'         // From search results (weight: 0.9)
  | 'playlist'       // From playlist (weight: 0.8)
  | 'artist'         // From artist page (weight: 0.75)
  | 'queue'          // From queue (weight: 0.7)
  | 'album'          // From album view (weight: 0.6)
  | 'recommendation' // From recommendations (weight: 0.7)
  | 'radio'          // From radio/discovery (weight: 0.4)
  | 'shuffle';       // From shuffle mode (weight: 0.2)

/**
 * Source type for play tracking
 */
export type SourceType = 'playlist' | 'album' | 'artist' | 'search' | 'queue' | 'recommendation';

/**
 * Play event data to send to backend
 */
export interface RecordPlayData {
  trackId: string;
  playContext: PlayContext;
  completionRate?: number;
  sourceId?: string;
  sourceType?: SourceType;
}

/**
 * Skip event data
 */
export interface RecordSkipData {
  trackId: string;
  timeListened: number;
  totalDuration: number;
  playContext: PlayContext;
  sourceId?: string;
  sourceType?: SourceType;
}

/**
 * Play history entry
 */
export interface PlayHistoryEntry {
  id: string;
  userId: string;
  trackId: string;
  playedAt: string;
  playContext: PlayContext;
  completionRate?: number;
  skipped: boolean;
  sourceId?: string;
  sourceType?: SourceType;
  track?: {
    id: string;
    title: string;
    artistName?: string;
    albumName?: string;
    duration?: number;
  };
}

/**
 * User play summary statistics
 */
export interface PlaySummary {
  totalPlays: number;
  uniqueTracks: number;
  totalListeningTime: number; // in seconds
  avgCompletionRate?: number;
  topPlayContext?: PlayContext;
  recentActivity: {
    last7Days: number;
    last30Days: number;
  };
}

/**
 * Top track with play stats
 */
export interface TopTrack {
  trackId: string;
  playCount: number;
  weightedPlayCount: number;
  avgCompletionRate?: number;
  skipCount: number;
  lastPlayedAt: string;
  track?: {
    id: string;
    title: string;
    artistName?: string;
    albumName?: string;
    duration?: number;
  };
}

/**
 * Recently played track
 */
export interface RecentlyPlayed {
  trackId: string;
  playedAt: string;
  playContext: PlayContext;
  completionRate?: number;
  track?: {
    id: string;
    title: string;
    artistName?: string;
    albumName?: string;
    duration?: number;
  };
}

/**
 * Record a play event
 * Called when a track completes or reaches significant playback
 */
export async function recordPlay(data: RecordPlayData): Promise<void> {
  try {
    await apiClient.post('/play-tracking/play', data);
  } catch (error) {
    if (import.meta.env.DEV) {
      logger.error('[PlayTracking] Failed to record play:', error);
    }
  }
}

/**
 * Record a skip event
 * Called when user skips to next track before completion
 */
export async function recordSkip(data: RecordSkipData): Promise<void> {
  try {
    const completionRate = data.totalDuration > 0
      ? data.timeListened / data.totalDuration
      : 0;

    await apiClient.post('/play-tracking/skip', {
      trackId: data.trackId,
      playContext: data.playContext,
      completionRate,
      sourceId: data.sourceId,
      sourceType: data.sourceType,
    });
  } catch (error) {
    if (import.meta.env.DEV) {
      logger.error('[PlayTracking] Failed to record skip:', error);
    }
  }
}

/**
 * Get user's play history
 */
export async function getPlayHistory(limit: number = 50): Promise<PlayHistoryEntry[]> {
  const response = await apiClient.get(`/play-tracking/history?limit=${limit}`);
  return response.data;
}

/**
 * Get user's top tracks
 */
export async function getTopTracks(
  limit: number = 20,
  timeRange: 'week' | 'month' | 'all' = 'all'
): Promise<TopTrack[]> {
  const response = await apiClient.get(`/play-tracking/top-tracks?limit=${limit}&timeRange=${timeRange}`);
  return response.data;
}

/**
 * Get recently played tracks
 */
export async function getRecentlyPlayed(limit: number = 20): Promise<RecentlyPlayed[]> {
  const response = await apiClient.get(`/play-tracking/recently-played?limit=${limit}`);
  return response.data;
}

/**
 * Get user's play summary statistics
 */
export async function getPlaySummary(): Promise<PlaySummary> {
  const response = await apiClient.get('/play-tracking/summary');
  return response.data;
}

/**
 * Playback state data for social "listening now" feature
 */
export interface PlaybackStateData {
  isPlaying: boolean;
  currentTrackId?: string | null;
}

/**
 * Update current playback state for social "listening now" feature
 * Called when playback starts, pauses, or stops
 */
export async function updatePlaybackState(data: PlaybackStateData): Promise<void> {
  try {
    await apiClient.put('/play-tracking/playback-state', data);
  } catch (error) {
    if (import.meta.env.DEV) {
      logger.error('[PlayTracking] Failed to update playback state:', error);
    }
  }
}
